### ISSUE TITLE
After upgrading "cucumber-picocontainer" from 1.2.5 to 5..0.0-RC1 getting "org.picocontainer.injectors.AbstractInjector$UnsatisfiableDependenciesException:"

### ISSUE THREAD

**Issue by LearningPur** [cucumber/cucumber-jvm#1822]
After upgrading pico-container version 5.0.0-RC1 from 1.2.5 i have started getting error message and desperately looking for help as our current production code is failing due to this migration. 

Any help asap would be appreciated.

**_Maven pom.xml change_** 

**Upgraded from this** 
            info.cukes
            cucumber-picocontainer
            1.2.5

**To this**
            io.cucumber
            cucumber-picocontainer
            5.0.0-RC1
            test
     
**_I am creating an instance of page class during runtime using these 4 list of classes_** 

**_Basepage.java_**
```
public class BasePage  {

    public static HomePage goToHomePage(Scenario scenario, String UserEmailID) {

        Selenide.refresh();
        if (WebDriverRunner.url().equalsIgnoreCase("https://www.footlocker.com/")) {
       }
        return page(HomePage.class);
    }
}
```
**BaseSteps.java**
```
import io.cucumber.java8.En;
import org.picocontainer.annotations.Inject;

public class BaseSteps implements En {

    @Inject
    private PageContainer pageContainer;

    protected void update(BasePage newPage) {
        pageContainer.updatePage(newPage);
    }

    protected BasePage page() {
        return pageContainer.getPage();
    }
}
```
**_PageContainer.java_**
```
public class PageContainer {

    private BasePage currentPage;

    public void updatePage(BasePage newPage) {
        currentPage = newPage;
    }

    public BasePage getPage() {
        return currentPage;
    }
}
```
**_PicoDependencyInjector.java_**
```
import io.cucumber.core.backend.ObjectFactory;
import io.cucumber.picocontainer.PicoFactory;

public class PicoDependencyInjector implements ObjectFactory {

    private PicoFactory delegate = new PicoFactory();

    public PicoDependencyInjector() {
        addClass(PageContainer.class);
    }


    @Override
    public void start() {
        delegate.start();
    }

    @Override
    public void stop() {
        delegate.stop();
    }

    @Override
    public boolean addClass(Class<?> aClass) {
        return delegate.addClass(aClass);
    }

    @Override
    public <T> T getInstance(Class<T> aClass) {
        return delegate.getInstance(aClass);
    }
}
```
**_cucumber.properties_**
```
io.cucumber.core.backend.ObjectFactory = com.footlocker.core.PicoDependencyInjector
```
_**Error i am getting is**_ 
```
org.picocontainer.injectors.AbstractInjector$UnsatisfiableDependenciesException: com.footlocker.steps.RegisterMultiUserSteps has unsatisfied dependency for fields [com.footlocker.core.PageContainer.pageContainer] from org.picocontainer.DefaultPicoContainer@661972b0:5<|

	at org.picocontainer.injectors.AbstractFieldInjector.unsatisfiedDependencies(AbstractFieldInjector.java:30)
	at org.picocontainer.injectors.IterativeInjector.getMatchingParameterListForSetters(IterativeInjector.java:104)
	at org.picocontainer.injectors.IterativeInjector.access$000(IterativeInjector.java:33)
	at org.picocontainer.injectors.IterativeInjector$2.run(IterativeInjector.java:153)
	at org.picocontainer.injectors.AbstractInjector$ThreadLocalCyclicDependencyGuard.observe(AbstractInjector.java:270)
	at org.picocontainer.injectors.IterativeInjector.getComponentInstance(IterativeInjector.java:160)
	at org.picocontainer.injectors.AbstractInjectionFactory$LifecycleAdapter.getComponentInstance(AbstractInjectionFactory.java:56)
	at org.picocontainer.behaviors.AbstractBehavior.getComponentInstance(AbstractBehavior.java:64)
	at org.picocontainer.behaviors.Stored.getComponentInstance(Stored.java:91)
	at org.picocontainer.DefaultPicoContainer.getInstance(DefaultPicoContainer.java:699)
	at org.picocontainer.DefaultPicoContainer.getComponent(DefaultPicoContainer.java:647)
	at org.picocontainer.DefaultPicoContainer.getComponent(DefaultPicoContainer.java:678)
	at io.cucumber.picocontainer.PicoFactory.getInstance(PicoFactory.java:47)
	at io.cucumber.java8.Java8Backend.buildWorld(Java8Backend.java:62)
	at io.cucumber.core.runner.Runner.buildBackendWorlds(Runner.java:165)
	at io.cucumber.core.runner.Runner.runPickle(Runner.java:61)
	at io.cucumber.junit.PickleRunners$NoStepDescriptions.run(PickleRunners.java:149)
	at io.cucumber.junit.FeatureRunner.runChild(FeatureRunner.java:83)
	at io.cucumber.junit.FeatureRunner.runChild(FeatureRunner.java:24)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at io.cucumber.junit.Cucumber.runChild(Cucumber.java:174)
	at io.cucumber.junit.Cucumber.runChild(Cucumber.java:82)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at io.cucumber.junit.Cucumber$RunCucumber.evaluate(Cucumber.java:208)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:230)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:58)

```

**Comment by mpkorstanje** [comment]
Heya, thanks for test driving RC1! 

We're using [Service Provider Interfaces](https://docs.oracle.com/javase/tutorial/sound/SPI-intro.html) to detect `ObjectFactory` implementations now. To make sure your object factory is picked up by SPI you'll have to create this file in your project:

`/src/main/resources/META-INF/services/io.cucumber.core.backend.ObjectFactory`
```
com.footlocker.core.PicoDependencyInjector
```

And your `cucumber.properties` file should contain:
```
cucumber.object-factory=com.footlocker.core.PicoDependencyInjector
```

Since you are here. Would you mind commenting on this issue https://github.com/cucumber/cucumber/issues/568? We're looking to create a list of companies that use Cucumber.

**Comment by LearningPur** [comment]
Thank you so much @mpkorstanje  for your quick reply

**Since you are here. Would you mind commenting on this issue cucumber/cucumber#568? We're looking to create a list of companies that use Cucumber.** 

I will for sure do that without forget

**We're using Service Provider Interfaces to detect ObjectFactory implementations now. To make sure your object factory is picked up by SPI you'll have to create this file in your project:
/src/main/resources/META-INF/services/io.cucumber.core.backend.ObjectFactory**

This is how i have added in src/main/resources if that is not correct way to do it, please do suggest as this is the first time i am doing this not sure how to get it there.

![Screen Shot 2019-11-17 at 3 07 12 PM](https://user-images.githubusercontent.com/57417403/69014089-48193f00-094c-11ea-8b34-23ce0133f5b5.png)

**And because you have more then one object factory on your class path, to disambiguate between the two your cucumber.properties file should contain:
cucumber.object-factory=com.footlocker.core.PicoDependencyInjector**

I have changed it as per your comment 

cucumber.object-factory=com.footlocker.core.PicoDependencyInjector

Thanks
Victor

**Comment by LearningPur** [comment]
@mpk

**Comment by mpkorstanje** [comment]
This means that your `cucumber.properties` file is correct but your service file is not. 

You have to create this file:

`/src/main/resources/META-INF/services/io.cucumber.core.backend.ObjectFactory` 

Note that the `/` are directory seperators and the `.` are part of the file name. The contents of the file should be:

```
com.footlocker.core.PicoDependencyInjector
```

Now I can't quite tell from your screen shot but it looks like you did not create a file but rather a directory. This is what it looks like for the `SpringObjectFactory`

![image](https://user-images.githubusercontent.com/6946919/69014313-2ee2b380-0989-11ea-9f52-e12392f9420e.png)

**Comment by LearningPur** [comment]
Yes, it worked. Thanks much.

**Comment by mpkorstanje** [comment]
I'm going to leave this open for a bit. There are a few things that should still be improved.

The Javadoc for `ObjectFactory` should mention that it is loaded via SPI. It currently doesn't. Infact it doesn't mention much at all.

> io.cucumber.core.exception.CucumberException: Could not find object factory class com.footlocker.core.PicoDependencyInjector

This exception should provide more information on what the exact problem is. The user selected an object factory but none were found. This means it either wasn't on the class path or the service wasn't registered correctly.

**Comment by mpkorstanje** [comment]
Resolved the doc issues in 0d864ca. The field scanning is too complicated for now. If some one wants to send a PR for it I'll take it though.
